r_code/synthetic_dgp.R

```diff
@@ -0,0 +1,1027 @@
+    dplyr::mutate(
+      cond_treat_impact = dplyr::case_when(
+        y0 > quantile(y0, 0.75) ~ cond_impact_inp,
+        y0 < quantile(y0, 0.25) ~ -cond_impact_inp,
+        TRUE ~ 0
+      )
+    ) %>%
+    dplyr::select(-y0)
+
+
+  # Define the treatment propagation for each unit and time combo.
+  data_inp <- data_inp %>%
+    dplyr::left_join(cond_x_walk, by = "entry") %>%
+    dplyr::group_by(entry) %>%
+    dplyr::mutate(
+      post_treat_t = time - treatment_period,
+      decay_t = dplyr::case_when(
+        post_treat_t < 0 ~ 0,
+        post_treat_t >= 0 ~ treat_decay**post_treat_t
+      ),
+      impact_t = decay_t * (treat_impact + cond_treat_impact),
+    ) %>%
+    dplyr::ungroup() %>%
+    dplyr::select(-c(treat_decay, treat_decay, decay_t, cond_treat_impact))
+
+  # Add the treatment impact to create potential treated outcome (y1).
+  # Also note the empirically observed y.
+  data_inp <- data_inp %>%
+    dplyr::mutate(
+      y1 = impact_t + y0,
+      y = treated * y1 + (1 - treated) * y0
+    ) %>%
+    dplyr::select(
+      time, entry, treated,
+      post_treat_t, treatment_period, impact_t,
+      y, y0, y1,
+      dplyr::everything()
+    )
+  return(data_inp)
+}
+
+#' Create a grid of date times.
+#'
+#' @param num_periods_inp Number of time periods to be generated.
+#' @param freq_inp String indicating time unit; "daily", "weekly", "monthly".
+#'
+#' @return Grid in freq_inp units spanning num_period_inp periods, with a
+#'    default start date of 2002-01-01.
+.GenerateTimeGrid <- function(num_periods_inp,
+                              freq_inp) {
+  # Creates a tibble that spans the input dates by the input frequency.
+  # Stores the relevant info about that date.
+  date_start_temp <- "2002-01-01"
+  period_dates <- switch(freq_inp,
+    "daily" = format(lubridate::ymd(date_start_temp) +
+      lubridate::days(0:(num_periods_inp - 1))),
+    "weekly" = format(lubridate::ymd(date_start_temp) +
+      lubridate::weeks(0:(num_periods_inp - 1))),
+    "monthly" = format(lubridate::ymd(date_start_temp) +
+      months(0:(num_periods_inp - 1)))
+  )
+
+  # Identify the relevant components of the date (day/week/etc).
+  date_info_tib <- tibble::tibble(
+    time = seq_len(num_periods_inp),
+    date_t = period_dates,
+    day_num = lubridate::day(period_dates),
+    yday_num = lubridate::yday(period_dates),
+    week_num = lubridate::week(period_dates),
+    month_num = lubridate::month(period_dates),
+    quarter_num = lubridate::quarter(period_dates),
+    year_num = lubridate::year(period_dates)
+  )
+
+  return(date_info_tib)
+}
+
+#' Creates the requisite number of factors beyond trend, given frequency.
+#'
+#' @param date_tib Grid of date time, spanning the desired number of periods.
+#' @param freq_inp String indicating time unit; "daily", "weekly", "monthly".
+#' @param shock_name String indicating the type of shock, one of "quarter_num",
+#'    "month_num", "week_num", "day_num".
```

Comment:
We should probably list the options in the parameter list and then use match.arg to choose one.

We should probably do this for any parameter in any function that actuals uses the values as opposed to passing them to function calls.  "Test at the leaves."  It may also make sense to test once at the top-level function as well.

Reply:
Is that redundant? The user inputs freq, which we match.arg at the top level. Then every helper function gets passed that match.arg'ed parameter. I don't have a strong sense of the proper way to do this, so I will follow your judgement!